### PR TITLE
Implement EIP-2537 using gnark-crypto

### PR DESCRIPTION
**Pull request by garyschulte**
### Implement [eip-2537](https://eips.ethereum.org/EIPS/eip-2537) using gnark-crypto.

related to #165 
related to #91 

* implements EIP-2537 using gnark-crypto, including as-spec'd subgroup checks and is-on-curve checks.  
* implements EIP-196 using gnark-crypto, adds test cases
* adds test cases for EIP-196 to matter-labs implementation
* shims gnark-crypto implementation to be a drop-in replacement for matter-labs 

### Motivation
The primary motivation is to get besu onto a supported library for the ECC precompiles since the matter-labs eip-1962 repo appears to be unmaintained.  There is additional benefit in that gnark-crypto performs better than matter labs on every precompile benchmark.

Testing on both x86_64 and ARM64 indicate this implementation compares very favorably with the matter-labs eip-1962 implementation for all operations.

Screenshot 2024-06-11 at 1 59 16 PM

Screenshot 2024-06-06 at 4 55 57 PM

Screenshot 2024-06-06 at 4 56 07 PM

Screenshot 2024-06-11 at 1 58 20 PM

Screenshot 2024-06-06 at 4 56 21 PM

Screenshot 2024-06-06 at 4 56 30 PM

Screenshot 2024-06-06 at 4 56 41 PM

Screenshot 2024-06-06 at 4 56 49 PM

Screenshot 2024-06-06 at 4 56 57 PM

Screenshot 2024-06-07 at 8 04 37 AM

Screenshot 2024-06-07 at 8 04 30 AM

Screenshot 2024-06-07 at 8 04 46 AM

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -16,10 +16,15 @@
 allprojects {
   apply plugin: 'java-library'
 
-  sourceCompatibility = 11
-  targetCompatibility = 11
+  sourceCompatibility = 17
+  targetCompatibility = 17
 
   repositories {
+    // temporary addition to enable local builds of v0.3.12 with g1/g2 mapping
```
Comment (garyschulte) [inline build.gradle:23]: remove

File: gnark/gnark-jni/gnark-eip-2537.go [hunk 2]
```diff
@@ -0,0 +1,550 @@
...
+    output := (*[EIP2537PreallocateForResultBytes]byte)(unsafe.Pointer(javaOutputBuf))[:outputLen:outputLen]
+
+    if inputLen != 2*EIP2537PreallocateForG1 {
+        copy(output, "invalid input parameters, invalid input length for G1 addition\x00")
+        return -1
+    }
+
+    // Convert input C pointers to Go slices
+    input := (*[2*EIP2537PreallocateForG1]byte)(unsafe.Pointer(javaInputBuf))[:inputLen:inputLen]
+//     fmt.Printf("convert input array time: %v\n", time.Since(startTime))
```
Comment (garyschulte) [inline gnark/gnark-jni/gnark-eip-2537.go:50]: remove

File: gnark/gnark-jni/gnark-eip-2537.go [hunk 3]
```diff
@@ -0,0 +1,550 @@
...
+        output[31]=0x01
+    }
+
+    return 0
+
+}
+
+//export eip2537blsMapFpToG1
+func eip2537blsMapFpToG1(javaInputBuf, javaOutputBuf *C.char, cInputLen, outputLen C.int) C.int {
+    //TODO: DRY up
```
Comment (garyschulte) [inline gnark/gnark-jni/gnark-eip-2537.go:345]: already pretty DRY

File: gnark/gnark-jni/gnark-eip-2537.go [hunk 4]
```diff
@@ -0,0 +1,550 @@
...
+
+    if inputLen != (EIP2537PreallocateForFp){
+        copy(output, "invalid input parameters, invalid input length for Fp to G1 to curve mapping\x00")
+        return -1
+    }
+
+    // Convert input C pointers to Go slice
+    input := (*[EIP2537PreallocateForFp]byte)(unsafe.Pointer(javaInputBuf))[:inputLen:inputLen]
+
+    // TODO: DRY up
```
Comment (garyschulte) [inline gnark/gnark-jni/gnark-eip-2537.go:357]: same

File: gnark/gnark-jni/gnark-eip-2537.go [hunk 5]
```diff
@@ -0,0 +1,550 @@
...
+    result := bls12381.MapToG1(fp)
+    // marshal the resulting point and encode directly to the output buffer
+    ret := result.Marshal()
+    g1AffineEncode(ret, javaOutputBuf)
+    return 0
+}
+
+//export eip2537blsMapFp2ToG2
+func eip2537blsMapFp2ToG2(javaInputBuf, javaOutputBuf *C.char, cInputLen, outputLen C.int) C.int {
+    //TODO: DRY up
```
Comment (garyschulte) [inline gnark/gnark-jni/gnark-eip-2537.go:370]: same

**Review by garyschulte — COMMENTED** [review]
fix a few minor things

File: build.gradle [hunk 6]
```diff
@@ -16,10 +16,13 @@
 allprojects {
   apply plugin: 'java-library'
 
-  sourceCompatibility = 11
-  targetCompatibility = 11
+  sourceCompatibility = 17
```
Comment (shemnon) [inline build.gradle:19]: 21?

**Review by shemnon — APPROVED** [review]
If gnark is the library of choice, should we make a plan to remove matterlabs from this repo?

File: build.gradle (hunk 6 above)
Comment (garyschulte) [reply build.gradle:19]: re: removing matter-labs, I was thinking we can remove it once we have a besu release with gnark, I have a PR ready to put up for that, it is just waiting for a release of besu-native that with this PR

Comment (garyschulte) [reply build.gradle]: thanks, made it 21 and bumped the project version to next minor rev to signify

**Comment by garyschulte** [comment]
> If gnark is the library of choice, should we make a plan to remove matterlabs from this repo?

Once we have a besu release with gnark-crypto, we can remove matter labs and the original altbn128 implementation 👍

**Review by shemnon — APPROVED** [review]

**Pull request merged**